### PR TITLE
GAWB-897: Workspace Last Modified Updates

### PR DESCRIPTION
**Pull request by rushtong**
This PR addresses [GAWB-897](https://broadinstitute.atlassian.net/browse/GAWB-897)

Added calls to update the workspace last modified date whenever any kind of modification operation is called. The following endpoints should all trigger an update:

**Admin:**
* DELETE /api/admin/submissions/{workspaceNamespace}/{workspaceName}/{submissionId}
~~* DELETE /api/admin/allUserReadAccess/{workspaceNamespace}/{workspaceName}~~
~~* PUT /api/admin/allUserReadAccess/{workspaceNamespace}/{workspaceName}~~

**Entities:**
* POST /api/workspaces/{workspaceNamespace}/{workspaceName}/entities/{entityType}/{entityName}/rename
* POST /api/workspaces/{workspaceNamespace}/{workspaceName}/entities
* PATCH /api/workspaces/{workspaceNamespace}/{workspaceName}/entities/{entityType}/{entityName}
* POST /api/workspaces/entities/copy
* POST /api/workspaces/{workspaceNamespace}/{workspaceName}/entities/batchUpsert
* POST /api/workspaces/{workspaceNamespace}/{workspaceName}/entities/batchUpdate

**Method Configs:**
* POST /api/workspaces/{workspaceNamespace}/{workspaceName}/methodconfigs
* DELETE /api/workspaces/{workspaceNamespace}/{workspaceName}/methodconfigs/{configNamespace}/{configName}
* PUT /api/workspaces/{workspaceNamespace}/{workspaceName}/methodconfigs/{configNamespace}/{configName}
* POST /api/workspaces/{workspaceNamespace}/{workspaceName}/methodconfigs/{configNamespace}/{configName}/rename
* POST /api/methodconfigs/copy

**Submissions:**
* DELETE /api/admin/submissions/{workspaceNamespace}/{workspaceName}/{submissionId}
* POST /api/workspaces/{workspaceNamespace}/{workspaceName}/submissions
* DELETE /api/workspaces/{workspaceNamespace}/{workspaceName}/submissions/{submissionId}

**Workspaces:**
~~* DELETE /api/admin/allUserReadAccess/{workspaceNamespace}/{workspaceName}~~
~~* PUT /api/admin/allUserReadAccess/{workspaceNamespace}/{workspaceName}~~
~~* PUT /api/workspaces/{workspaceNamespace}/{workspaceName}/lock~~
~~* PUT /api/workspaces/{workspaceNamespace}/{workspaceName}/unlock~~
~~* PATCH /api/workspaces/{workspaceNamespace}/{workspaceName}/acl~~ 
* POST /api/workspaces 
* PATCH /api/workspaces/{workspaceNamespace}/{workspaceName} 

---
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell your tech lead (TL) that the PR exists if they want to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by dvoet** [comment]
Have you been too far reaching? I question whether access control changes, workspace lock/unlock, submission creation/termination/completion should be included. The ticket does not really mention any of these things. I would expect the updated date to change only when "content" of the workspace changes. Content to me includes only entity/workspace attributes and method configs. That is how I read the ticket.

**Comment by rushtong** [comment]
@dvoet - See the parent story: https://broadinstitute.atlassian.net/browse/GAWB-1019
Alex wants anything that changes inside a workspace to change the last updated date.

**Comment by dvoet** [comment]
I did, and my point stands. I think you should double check with the POs the exact list of things that will cause an update and make sure they are fine with it.

**Comment by rushtong** [comment]
Based on feedback from Alex, I'll update this and remove the updates on these endpoints:
- DELETE /api/admin/allUserReadAccess/{workspaceNamespace}/{workspaceName}
- PUT /api/admin/allUserReadAccess/{workspaceNamespace}/{workspaceName}
- DELETE /api/admin/workspaces/{workspaceNamespace}/{workspaceName} 
- PUT /api/workspaces/{workspaceNamespace}/{workspaceName}/lock 
- PUT /api/workspaces/{workspaceNamespace}/{workspaceName}/unlock 
- PATCH /api/workspaces/{workspaceNamespace}/{workspaceName}/acl

**Comment by davidangb** [comment]
fwiw, I also negotiated this ticket with Alex before this sprint, and he confirmed to me too that he wants ANY change to update the lastmodified date. Glad you got answers, and sorry you had to go through that - thought it was resolved. Doug, I had the same skepticism, but Alex was certain.

regarding the code:
- is this PR adding a new `UPDATE` SQL query for all these operations? Not sure there's a way around this, but I want to understand it.
- I see in some cases you update the lastModifiedDate, then perform the original action; in other cases you perform the original action, then update lastModifiedDate. Any reason for this?
- do any of these need transactions?

now that you've done all of this work, could/should any of this be done with database triggers?

File: src/test/scala/org/broadinstitute/dsde/rawls/WorkspaceTestUtils.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  def assertWorkspaceResult(expected: Seq[Workspace])(actual: Seq[Workspace]) = {
+    expected zip actual foreach {
+      case (e, a) =>
+        assertResult(e) {
+          a.copy(lastModified = e.lastModified)
+        }
+      case _ => throw new TestFailedException(Some("Unable to parse expected value: " + expected.toString), None, 4)
+    }
+  }
```
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/rawls/WorkspaceTestUtils.scala:51]: could you add comments explaining why we need these `assertWorkspaceResult` methods? I understand why, but upon first reading it, I was alarmed that you're munging the expected object.

**Comment by rushtong** [comment]
@davidangb 

> is this PR adding a new UPDATE SQL query for all these operations? Not sure there's a way around this, but I want to understand it.

Yes, that is true - I don't see any way around this (aside from triggers)

> I see in some cases you update the lastModifiedDate, then perform the original action; in other cases you perform the original action, then update lastModifiedDate. Any reason for this?

In some cases, changing the order changes the return of the method signature. I tried to avoid that as much as possible. In all cases, the updates are inside a transaction - so the order should be irrelevant as long as they all happen together. 

> do any of these need transactions?

Yes - they are all inside of a transaction.

**Comment by dvoet** [comment]
I have also been trying to think of alternatives to adding an update everywhere. What I don't like about the current approach is that it is kind of side effecty (i.e. save an attribute also updates the workspace record) and we may be prone to forget to do this when we add new features.

But I don't have an alternative that I like better.

File: src/test/scala/org/broadinstitute/dsde/rawls/WorkspaceTestUtils.scala (hunk 1 above)
Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/rawls/WorkspaceTestUtils.scala:51]: ✔️

**Comment by rushtong** [comment]
@dvoet - Any further feedback on this? I agree with the side-affecty-ness of this, but I could not come up with a better approach.

**Comment by dvoet** [comment]
no, proceed

**Pull request merged**